File: school_timetable/app/services/conflict_checker.py
"""
Comprehensive Conflict Checker Service.
Audits timetables against hard constraints and soft constraints:
- Teacher double booking
- Class double booking
- Teacher availability violations
- Teacher qualification mismatches
- Teacher daily overload
- Teacher weekly overload
- Missing or extra subject periods
- Break/lunch period overlaps
"""

from typing import List, Dict, Any, Optional
from collections import defaultdict
from app.schemas import ConflictItem


class ConflictChecker:
    @staticmethod
    def audit_timetable(
        entries: List[Dict[str, Any]],
        classes: List[Dict[str, Any]],
        staff_list: List[Dict[str, Any]],
        subjects: List[Dict[str, Any]],
        assignments: List[Dict[str, Any]],
        availabilities: List[Dict[str, Any]],
        class_subjects: List[Dict[str, Any]],
        active_days: List[str],
        total_periods: int = 8,
        lunch_period: Optional[int] = None,
        break_periods: Optional[List[int]] = None,
    ) -> List[ConflictItem]:
        conflicts: List[ConflictItem] = []
        break_periods = break_periods or []

        # Maps for quick lookups
        class_map = {c["id"]: c["name"] for c in classes}
        staff_map = {s["id"]: s for s in staff_list}
        subject_map = {sub["id"]: sub for sub in subjects}

        # 1. HARD CONSTRAINT: Teacher Double Booking
        # teacher_id + day + period must be unique
        teacher_slots: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
        for e in entries:
            key = f"{e['staff_id']}_{e['day']}_{e['period']}"
            teacher_slots[key].append(e)

        for key, slot_entries in teacher_slots.items():
            if len(slot_entries) > 1:
                st_id = slot_entries[0]["staff_id"]
                st_name = staff_map.get(st_id, {}).get("name", f"Teacher #{st_id}")
                day = slot_entries[0]["day"]
                period = slot_entries[0]["period"]
                cls_names = [class_map.get(x["class_id"], f"Class {x['class_id']}") for x in slot_entries]
                sub_names = [subject_map.get(x["subject_id"], {}).get("name", "Subject") for x in slot_entries]

                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="teacher_double_booking",
                    day=day,
                    period=period,
                    staff_name=st_name,
                    class_name=", ".join(cls_names),
                    subject_name=", ".join(sub_names),
                    explanation=f"Teacher {st_name} is simultaneously double-booked to teach classes ({', '.join(cls_names)}) on {day} Period {period}.",
                    suggested_fix=f"Reschedule one of the classes to a free period for {st_name} or assign an alternate qualified teacher."
                ))

        # 2. HARD CONSTRAINT: Class Double Booking
        # class_id + day + period must be unique
        class_slots: Dict[str, List[Dict[str, Any]]] = defaultdict(list)
        for e in entries:
            key = f"{e['class_id']}_{e['day']}_{e['period']}"
            class_slots[key].append(e)

        for key, slot_entries in class_slots.items():
            if len(slot_entries) > 1:
                c_id = slot_entries[0]["class_id"]
                c_name = class_map.get(c_id, f"Class #{c_id}")
                day = slot_entries[0]["day"]
                period = slot_entries[0]["period"]
                sub_names = [subject_map.get(x["subject_id"], {}).get("name", "Subject") for x in slot_entries]
                st_names = [staff_map.get(x["staff_id"], {}).get("name", "Teacher") for x in slot_entries]

                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="class_double_booking",
                    day=day,
                    period=period,
                    class_name=c_name,
                    staff_name=", ".join(st_names),
                    subject_name=", ".join(sub_names),
                    explanation=f"Class {c_name} has multiple subjects ({', '.join(sub_names)}) scheduled at the same time on {day} Period {period}.",
                    suggested_fix=f"Move one subject to an open period on {day} or another active weekday."
                ))

        # 3. HARD CONSTRAINT: Teacher Availability
        unavail_set = set()
        for a in availabilities:
            if not a.get("is_available", True):
                unavail_set.add((a["staff_id"], a["day"], a["period"]))

        for e in entries:
            if (e["staff_id"], e["day"], e["period"]) in unavail_set:
                st_name = staff_map.get(e["staff_id"], {}).get("name", f"Teacher #{e['staff_id']}")
                c_name = class_map.get(e["class_id"], f"Class #{e['class_id']}")
                sub_name = subject_map.get(e["subject_id"], {}).get("name", "Subject")
                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="teacher_unavailable",
                    day=e["day"],
                    period=e["period"],
                    class_name=c_name,
                    staff_name=st_name,
                    subject_name=sub_name,
                    explanation=f"{st_name} is marked unavailable on {e['day']} Period {e['period']}, but is assigned to teach {c_name} ({sub_name}).",
                    suggested_fix=f"Move this lesson to when {st_name} is available, or adjust {st_name}'s availability preferences."
                ))

        # 4. HARD CONSTRAINT: Teacher Qualification
        # Teacher must be qualified for subject
        staff_qual_map: Dict[int, set] = defaultdict(set)
        for s in staff_list:
            for sub_id in s.get("qualified_subject_ids", []):
                staff_qual_map[s["id"]].add(sub_id)

        for e in entries:
            st_id = e["staff_id"]
            sub_id = e["subject_id"]
            if sub_id not in staff_qual_map.get(st_id, set()):
                st_name = staff_map.get(st_id, {}).get("name", f"Teacher #{st_id}")
                c_name = class_map.get(e["class_id"], f"Class #{e['class_id']}")
                sub_name = subject_map.get(sub_id, {}).get("name", f"Subject #{sub_id}")
                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="qualification_mismatch",
                    day=e["day"],
                    period=e["period"],
                    class_name=c_name,
                    staff_name=st_name,
                    subject_name=sub_name,
                    explanation=f"{st_name} is teaching {sub_name} for {c_name} on {e['day']} Period {e['period']}, but is not listed as qualified for {sub_name}.",
                    suggested_fix=f"Update {st_name}'s qualified subjects or assign a qualified teacher."
                ))

        # 5. HARD CONSTRAINT: Teacher Daily & Weekly Overload
        teacher_daily_counts: Dict[str, int] = defaultdict(int)
        teacher_weekly_counts: Dict[int, int] = defaultdict(int)

        for e in entries:
            st_id = e["staff_id"]
            day = e["day"]
            teacher_daily_counts[f"{st_id}_{day}"] += 1
            teacher_weekly_counts[st_id] += 1

        for st_id, staff_obj in staff_map.items():
            max_day = staff_obj.get("max_periods_per_day", 6)
            max_week = staff_obj.get("max_periods_per_week", 30)

            for d in active_days:
                count = teacher_daily_counts.get(f"{st_id}_{d}", 0)
                if count > max_day:
                    conflicts.append(ConflictItem(
                        severity="medium",
                        conflict_type="daily_overload",
                        day=d,
                        staff_name=staff_obj["name"],
                        explanation=f"{staff_obj['name']} is assigned {count} periods on {d}, exceeding their maximum limit of {max_day} periods/day.",
                        suggested_fix=f"Shift some periods from {d} to another day with lighter workload."
                    ))

            week_count = teacher_weekly_counts.get(st_id, 0)
            if week_count > max_week:
                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="weekly_overload",
                    staff_name=staff_obj["name"],
                    explanation=f"{staff_obj['name']} has {week_count} total periods assigned this week, exceeding the weekly maximum of {max_week}.",
                    suggested_fix=f"Reassign some classes/periods to another staff member."
                ))

        # 6. HARD CONSTRAINT: Subject Weekly Requirements (Missing / Extra)
        required_map: Dict[str, int] = {}
        for cs in class_subjects:
            required_map[f"{cs['class_id']}_{cs['subject_id']}"] = cs["periods_per_week"]

        scheduled_counts: Dict[str, int] = defaultdict(int)
        for e in entries:
            scheduled_counts[f"{e['class_id']}_{e['subject_id']}"] += 1

        for key, req_count in required_map.items():
            c_id_str, s_id_str = key.split("_")
            c_id, s_id = int(c_id_str), int(s_id_str)
            actual_count = scheduled_counts.get(key, 0)
            c_name = class_map.get(c_id, f"Class #{c_id}")
            sub_name = subject_map.get(s_id, {}).get("name", f"Subject #{s_id}")

            if actual_count < req_count:
                missing = req_count - actual_count
                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="missing_subject_periods",
                    class_name=c_name,
                    subject_name=sub_name,
                    explanation=f"{c_name} requires {req_count} periods/week of {sub_name}, but only {actual_count} are scheduled (Missing: {missing}).",
                    suggested_fix=f"Generate or assign {missing} more period(s) for {sub_name} in {c_name}."
                ))
            elif actual_count > req_count:
                extra = actual_count - req_count
                conflicts.append(ConflictItem(
                    severity="medium",
                    conflict_type="extra_subject_periods",
                    class_name=c_name,
                    subject_name=sub_name,
                    explanation=f"{c_name} has {actual_count} periods of {sub_name} scheduled, exceeding the required {req_count} by {extra} period(s).",
                    suggested_fix=f"Remove {extra} extra period(s) of {sub_name} from {c_name}'s schedule."
                ))

        # 7. HARD CONSTRAINT: Break and Lunch overlaps
        for e in entries:
            p = e["period"]
            if lunch_period is not None and p == lunch_period:
                c_name = class_map.get(e["class_id"], f"Class #{e['class_id']}")
                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="lunch_overlap",
                    day=e["day"],
                    period=p,
                    class_name=c_name,
                    explanation=f"Class {c_name} is scheduled during designated Lunch time (Period {p}).",
                    suggested_fix="Move class to an academic period."
                ))
            if p in break_periods:
                c_name = class_map.get(e["class_id"], f"Class #{e['class_id']}")
                conflicts.append(ConflictItem(
                    severity="high",
                    conflict_type="break_overlap",
                    day=e["day"],
                    period=p,
                    class_name=c_name,
                    explanation=f"Class {c_name} is scheduled during designated Break period (Period {p}).",
                    suggested_fix="Move class to an instructional period."
                ))

        return conflicts

File: src/services/conflictChecker.ts
import {
  TimetableEntry,
  SchoolClass,
  Staff,
  Subject,
  StaffAssignment,
  SchoolTimings,
  ConflictItem,
} from "../types";

export class ConflictChecker {
  static audit(
    entries: TimetableEntry[],
    classes: SchoolClass[],
    staffList: Staff[],
    subjects: Subject[],
    assignments: StaffAssignment[],
    timings: SchoolTimings,
    lunchPeriod = 5,
    breakPeriods: number[] = [3, 7]
  ): ConflictItem[] {
    const conflicts: ConflictItem[] = [];

    const classMap = new Map(classes.map((c) => [c.id, c.name]));
    const staffMap = new Map(staffList.map((s) => [s.id, s]));
    const subjectMap = new Map(subjects.map((s) => [s.id, s]));

    // 1. HARD CONSTRAINT: Teacher Double Booking
    const teacherSlots = new Map<string, TimetableEntry[]>();
    for (const e of entries) {
      const key = `${e.staff_id}_${e.day}_${e.period}`;
      if (!teacherSlots.has(key)) teacherSlots.set(key, []);
      teacherSlots.get(key)!.push(e);
    }

    teacherSlots.forEach((slotEntries) => {
      if (slotEntries.length > 1) {
        const st = staffMap.get(slotEntries[0].staff_id);
        const day = slotEntries[0].day;
        const period = slotEntries[0].period;
        const classNames = slotEntries.map((e) => classMap.get(e.class_id) || `Class ${e.class_id}`);
        const subjectNames = slotEntries.map((e) => subjectMap.get(e.subject_id)?.name || "Subject");

        conflicts.push({
          severity: "high",
          conflict_type: "teacher_double_booking",
          day,
          period,
          staff_name: st?.name || "Teacher",
          class_name: classNames.join(", "),
          subject_name: subjectNames.join(", "),
          explanation: `Teacher ${st?.name || "Teacher"} is simultaneously scheduled to teach classes (${classNames.join(", ")}) on ${day} Period ${period}.`,
          suggested_fix: `Move one of the classes to a free period for ${st?.name || "Teacher"}.`,
        });
      }
    });

    // 2. HARD CONSTRAINT: Class Double Booking
    const classSlots = new Map<string, TimetableEntry[]>();
    for (const e of entries) {
      const key = `${e.class_id}_${e.day}_${e.period}`;
      if (!classSlots.has(key)) classSlots.set(key, []);
      classSlots.get(key)!.push(e);
    }

    classSlots.forEach((slotEntries) => {
      if (slotEntries.length > 1) {
        const cName = classMap.get(slotEntries[0].class_id) || "Class";
        const day = slotEntries[0].day;
        const period = slotEntries[0].period;
        const subNames = slotEntries.map((e) => subjectMap.get(e.subject_id)?.name || "Subject");
        const staffNames = slotEntries.map((e) => staffMap.get(e.staff_id)?.name || "Teacher");

        conflicts.push({
          severity: "high",
          conflict_type: "class_double_booking",
          day,
          period,
          class_name: cName,
          staff_name: staffNames.join(", "),
          subject_name: subNames.join(", "),
          explanation: `Class ${cName} has multiple subjects (${subNames.join(", ")}) scheduled at the same time on ${day} Period ${period}.`,
          suggested_fix: `Move one subject to an open period on ${day} or another weekday.`,
        });
      }
    });

    // 3. HARD CONSTRAINT: Teacher Availability
    const unavailSet = new Set<string>();
    for (const s of staffList) {
      for (const u of s.unavailabilities || []) {
        unavailSet.add(`${s.id}_${u.day}_${u.period}`);
      }
    }

    for (const e of entries) {
      if (unavailSet.has(`${e.staff_id}_${e.day}_${e.period}`)) {
        const stName = staffMap.get(e.staff_id)?.name || "Teacher";
        const cName = classMap.get(e.class_id) || "Class";
        const subName = subjectMap.get(e.subject_id)?.name || "Subject";
        conflicts.push({
          severity: "high",
          conflict_type: "teacher_unavailable",
          day: e.day,
          period: e.period,
          class_name: cName,
          staff_name: stName,
          subject_name: subName,
          explanation: `${stName} is marked unavailable on ${e.day} Period ${e.period}, but is assigned to teach ${cName} (${subName}).`,
          suggested_fix: `Move this lesson to when ${stName} is available, or adjust availability.`,
        });
      }
    }

    // 4. HARD CONSTRAINT: Teacher Qualifications & Authorizations
    for (const e of entries) {
      const st = staffMap.get(e.staff_id);
      const isAssigned = assignments.some(
        (a) => a.staff_id === e.staff_id && a.subject_id === e.subject_id && a.class_id === e.class_id
      );
      if (st && !isAssigned && !st.qualified_subject_ids.includes(e.subject_id)) {
        const subName = subjectMap.get(e.subject_id)?.name || "Subject";
        conflicts.push({
          severity: "high",
          conflict_type: "qualification_mismatch",
          day: e.day,
          period: e.period,
          class_name: classMap.get(e.class_id) || "Class",
          staff_name: st.name,
          subject_name: subName,
          explanation: `${st.name} is scheduled to teach ${subName}, but is neither assigned nor qualified for it.`,
          suggested_fix: `Assign ${st.name} in Staff Assignments or select another teacher.`,
        });
      }
    }

    // 5. HARD CONSTRAINT: Daily and Weekly limits
    const dailyCount = new Map<string, number>();
    const weeklyCount = new Map<number, number>();

    for (const e of entries) {
      const dayKey = `${e.staff_id}_${e.day}`;
      dailyCount.set(dayKey, (dailyCount.get(dayKey) || 0) + 1);
      weeklyCount.set(e.staff_id, (weeklyCount.get(e.staff_id) || 0) + 1);
    }

    for (const st of staffList) {
      for (const d of timings.active_days) {
        const count = dailyCount.get(`${st.id}_${d}`) || 0;
        if (count > st.max_periods_per_day) {
          conflicts.push({
            severity: "medium",
            conflict_type: "daily_overload",
            day: d,
            staff_name: st.name,
            explanation: `${st.name} has ${count} periods assigned on ${d} (Limit: ${st.max_periods_per_day}/day).`,
            suggested_fix: `Move some periods to other active days with lighter workload.`,
          });
        }
      }

      const totalWeek = weeklyCount.get(st.id) || 0;
      const assignedCount = assignments
        .filter((a) => a.staff_id === st.id)
        .reduce((sum, a) => {
          const cls = classes.find((c) => c.id === a.class_id);
          const req = cls?.subjects.find((s) => s.subject_id === a.subject_id);
          return sum + (req?.periods_per_week || 0);
        }, 0);
      const effectiveMaxWeek = Math.max(st.max_periods_per_week, assignedCount);
      const maxSlotsInWeek = timings.active_days.length * timings.total_periods;

      if (totalWeek > effectiveMaxWeek || totalWeek > maxSlotsInWeek) {
        conflicts.push({
          severity: "high",
          conflict_type: "weekly_overload",
          staff_name: st.name,
          explanation: `${st.name} has ${totalWeek} total periods this week (Limit: ${effectiveMaxWeek}/week).`,
          suggested_fix: `Reassign classes/periods to another teacher.`,
        });
      }
    }

    // 6. Subject requirements per class (missing or extra)
    const requiredMap = new Map<string, number>();
    for (const c of classes) {
      for (const req of c.subjects) {
        requiredMap.set(`${c.id}_${req.subject_id}`, req.periods_per_week);
      }
    }

    const scheduledMap = new Map<string, number>();
    for (const e of entries) {
      const key = `${e.class_id}_${e.subject_id}`;
      scheduledMap.set(key, (scheduledMap.get(key) || 0) + 1);
    }

    requiredMap.forEach((reqCount, key) => {
      const [cIdStr, sIdStr] = (key || "").split("_");
      const cId = parseInt(cIdStr, 10);
      const sId = parseInt(sIdStr, 10);
      const scheduled = scheduledMap.get(key) || 0;
      const cName = classMap.get(cId) || "Class";
      const sName = subjectMap.get(sId)?.name || "Subject";

      if (scheduled < reqCount) {
        conflicts.push({
          severity: "high",
          conflict_type: "missing_subject_periods",
          class_name: cName,
          subject_name: sName,
          explanation: `${cName} requires ${reqCount} periods/week of ${sName}, but only ${scheduled} are assigned (Missing: ${reqCount - scheduled}).`,
          suggested_fix: `Generate or schedule ${reqCount - scheduled} additional period(s).`,
        });
      } else if (scheduled > reqCount) {
        conflicts.push({
          severity: "medium",
          conflict_type: "extra_subject_periods",
          class_name: cName,
          subject_name: sName,
          explanation: `${cName} has ${scheduled} periods of ${sName} scheduled, exceeding required ${reqCount} by ${scheduled - reqCount}.`,
          suggested_fix: `Remove ${scheduled - reqCount} extra period(s).`,
        });
      }
    });

    return conflicts;
  }

  static canAssignSlot(
    entries: TimetableEntry[],
    classId: number,
    staffId: number,
    day: string,
    period: number,
    timings: SchoolTimings,
    staffList: Staff[],
    excludeEntryId?: number
  ): { allowed: boolean; reason?: string } {
    // 1. Period bounds check
    if (period < 1 || period > timings.total_periods) {
      return { allowed: false, reason: `Period ${period} is outside valid range (1 to ${timings.total_periods}).` };
    }

    // Filter out the entry being moved
    const otherEntries = entries.filter((e) => e.id !== excludeEntryId);

    // 2. Check Teacher collision
    const teacherClash = otherEntries.find(
      (e) => e.staff_id === staffId && e.day === day && e.period === period
    );
    if (teacherClash) {
      return {
        allowed: false,
        reason: `Teacher is already teaching Class ${teacherClash.class_id} on ${day} Period ${period}.`,
      };
    }

    // 3. Check Class collision
    const classClash = otherEntries.find(
      (e) => e.class_id === classId && e.day === day && e.period === period
    );
    if (classClash) {
      return {
        allowed: false,
        reason: `Class is already scheduled for another lesson on ${day} Period ${period}.`,
      };
    }

    // 4. Check Teacher unavailability
    const staff = staffList.find((s) => s.id === staffId);
    if (staff) {
      const isBlocked = (staff.unavailabilities || []).some(
        (u) => u.day === day && u.period === period
      );
      if (isBlocked) {
        return {
          allowed: false,
          reason: `Teacher ${staff.name} is marked as unavailable (Meeting / Leave) on ${day} Period ${period}.`,
        };
      }

      // 5. Daily limit
      const dailyCount = otherEntries.filter(
        (e) => e.staff_id === staffId && e.day === day
      ).length;
      if (dailyCount >= staff.max_periods_per_day) {
        return {
          allowed: false,
          reason: `Teacher ${staff.name} would exceed max daily teaching limit of ${staff.max_periods_per_day} periods on ${day}.`,
        };
      }
    }

    return { allowed: true };
  }
}
